h/api/parse_document_claims.py

```diff
@@ -0,0 +1,215 @@
+    Return one or more document URI dicts for the given document data.
+
+    Returns one document uri dict for each document equivalence claim in
+    document_data.
+
+    Each dict can be used to init a DocumentURI object directly:
+
+        document_uri = DocumentURI(**document_uri_dict)
+
+    Always returns at least one "self-claim" document URI whose URI is the
+    claimant URI itself.
+
+    :param document_data: the "document" sub-object that was POSTed to the API
+        as part of a new or updated annotation
+    :type document_data: dict
+
+    :param claimant: the URI that the browser was at when this annotation was
+        created (the top-level "uri" field of the annotation)
+    :type claimant: unicode
+
+    :returns: a list of one or more document URI dicts
+    :rtype: list of dicts
+
+    """
+    document_uris = document_uris_from_links(document_data.get('link', []),
+                                             claimant)
+
+    document_uris.extend(
+        document_uris_from_highwire(document_data.get('highwire', {}),
+                                    claimant)
+    )
+
+    document_uris.extend(
+        document_uris_from_dc(document_data.get('dc', {}),
+                              claimant)
+    )
+
+    document_uris.append(document_uri_self_claim(claimant))
+
+    return document_uris
+
+
+def document_metas_from_data(document_data, claimant):
+    """
+    Return a list of document meta dicts for the given document data.
+
+    Returns one document meta dict for each document metadata claim in
+    document_data.
+
+    Each dict can be used to init a DocumentMeta object directly:
+
+        document_meta = DocumentMeta(**document_meta_dict)
+
+    :param document_data: the "document" sub-object that the client POSTed to
+        the API as part of a new or updated annotation
+    :type document_data: dict
+
+    :param claimant: the URI that the browser was at when this annotation was
+        created (the top-level "uri" field of the annotation)
+    :type claimant: unicode
+
+    :returns: a list of zero or more document meta dicts
+    :rtype: list of dicts
+
+    """
+    def transform_meta_(document_meta_dicts, items, path_prefix=None):
+        """Fill document_meta_dicts with document meta dicts for the items."""
+        if path_prefix is None:
+            path_prefix = []
+
+        for key, value in items.iteritems():
+            keypath = path_prefix[:]
+            keypath.append(key)
+
+            if isinstance(value, dict):
+                transform_meta_(document_meta_dicts,
+                                value,
+                                path_prefix=keypath)
+            else:
+                if not isinstance(value, list):
+                    value = [value]
+
+                document_meta_dicts.append({
+                    'type': '.'.join(keypath),
+                    'value': value,
+                    'claimant': claimant,
+                    'claimant_normalized':
```

Comment:
There's no need to set the `claimant_normalized` manually, the `claimant` setter takes care of this.